services/gossip/adapter/tcp/direct_outgoing.go

```diff
@@ -10,56 +10,107 @@ import (
+	peerAddress := fmt.Sprintf("%s:%d", peer.GossipEndpoint(), peer.GossipPort())
+	queue := NewTransportQueue(SEND_QUEUE_MAX_BYTES, SEND_QUEUE_MAX_MESSAGES, metricFactory, peerHexAddress)
+	queue.networkAddress = peerAddress
+	queue.Disable() // until connection is established
+
+	client := &clientConnection{
+		logger:          parentLogger.WithTags(log.String("peer-node-address", peerHexAddress), log.String("peer-network-address", peerAddress)),
+		sharedMetrics:   sharedMetrics,
+		config:          transportConfig,
+		queue:           queue,
+		sendErrors:      metricFactory.NewGauge(fmt.Sprintf("Gossip.OutgoingConnection.SendError.%s.Count", peerHexAddress)),
+		sendQueueErrors: metricFactory.NewGauge(fmt.Sprintf("Gossip.OutgoingConnection.EnqueueErrors.%s.Count", peerHexAddress)),
+	}
+
+	return client
+}
+
+func (c *clientConnection) connect(parent context.Context) {
+	ctx, cancel := context.WithCancel(parent)
+	c.disconnect = cancel
+
+	supervised.GoForever(ctx, c.logger, func() {
+		c.clientMainLoop(ctx, c.queue) // avoid referencing queue map not under lock
+	})
+}
+
+func (c *clientConnection) clientMainLoop(parentCtx context.Context, queue *transportQueue) {
 	for {
-		ctx := trace.NewContext(parentCtx, fmt.Sprintf("Gossip.Transport.TCP.Client.%s", queue.networkAddress))
-		t.logger.Info("attempting outgoing transport connection", log.String("peer", queue.networkAddress), trace.LogFieldFrom(ctx))
-		conn, err := net.DialTimeout("tcp", queue.networkAddress, t.config.GossipNetworkTimeout())
+		ctx := trace.NewContext(parentCtx, fmt.Sprintf("Gossip.Transport.TCP.Client.%s", c.peerNodeAddress))
+		logger := c.logger.WithTags(trace.LogFieldFrom(ctx))
+
+		logger.Info("attempting outgoing transport connection")
+		conn, err := net.DialTimeout("tcp", queue.networkAddress, c.config.GossipNetworkTimeout())
 
 		if err != nil {
-			t.logger.Info("cannot connect to gossip peer endpoint", log.String("peer", queue.networkAddress), trace.LogFieldFrom(ctx))
-			time.Sleep(t.config.GossipReconnectInterval())
+			logger.Info("cannot connect to gossip peer endpoint")
+			time.Sleep(c.config.GossipReconnectInterval())
 			continue
 		}
 
-		if !t.clientHandleOutgoingConnection(ctx, conn, queue) {
+		if !c.clientHandleOutgoingConnection(ctx, conn, queue) {
 			return
 		}
 	}
 }
 
 // returns true if should attempt reconnect on error
-func (t *DirectTransport) clientHandleOutgoingConnection(ctx context.Context, conn net.Conn, queue *transportQueue) bool {
-	t.logger.Info("successful outgoing gossip transport connection", log.String("peer", queue.networkAddress), trace.LogFieldFrom(ctx))
-	t.metrics.activeOutgoingConnections.Inc()
-	defer t.metrics.activeOutgoingConnections.Dec()
+func (c *clientConnection) clientHandleOutgoingConnection(ctx context.Context, conn net.Conn, queue *transportQueue) bool {
```

Comment:
suggest to extract some of the branching code here into functions - replace comments with function names to make this easier to read